File: modules_js/blocks.js
module.exports = {
    hand_shaking_server: function () {
        var tools = require('../modules_js/modules.js');
        
        /*  espera o primeiro ACK */
        do {
            data = tools.read_file('transporte+fisica.txt');
            array_data = data.split(tools.control_separator);
        } while ((array_data[1] == 'TRANSPORT_DIDNT|PHYSICAL_DIDNT')
            || (array_data[1] == 'TRANSPORT_DONE|PHYSICAL_DIDNT'));


        /* temos que fragmentar o conteudo para atualizar o ack e o seq*/
        var fragmented_data = array_data[0].split(':');
        /* atualizamos o ACK e o SEQ */
        fragmented_data[2] = fragmented_data[2] + 1;
        fragmented_data[3] = fragmented_data[3] + fragmented_data[0].length;
        array_data[0] = fragmented_data.join(tools.atribute_separator);
        data = array_data[0] + '=TRANSPORT_DONE|PHYSICAL_DIDNT';

        // escreve segundo ACK no arquivo
        tools.write_file('transporte+fisica.txt', data);

        do {
            /* espera o terceiro ACK */
            data = tools.read_file('transporte+fisica.txt');
            array_data = data.split(tools.control_separator);
        } while (array_data[1] == '=TRANSPORT_DONE|PHYSICAL_DIDNT');
    },

    hand_shaking_client: function(source_port,destination_port) {

        var tools = require('../modules_js/modules.js');

        var seq = 1;
        var ack = 0;

        // primeiro ack - indexa a PDU da camada de transporte
        var data = tools.create_pdu('3_WAY_HANDSHAKE', seq, ack, source_port, destination_port);
        var array_data = data.split(tools.atributes_separator);
        // atualiza o ack
        ack = array_data[3];

        data += '=TRANSPORT_DONE|PHYSICAL_DIDNT';

        // escreve o primeiro ack no arquivo 
        tools.write_file('transporte+fisica.txt', data);

        do {
            /* espera o segundo ack do servidor */
            data = tools.read_file('transporte+fisica.txt');
            array_data = data.split('=');
        } while (array_data[1] == 'TRANSPORT_DONE|PHYSICAL_DIDNT');

        /* temos que fragmentar o conteudo para atualizar o ack e o seq*/
        var fragmented_data = array_data[0].split(tools.atributes_separator);
        /* atualizamos o SEQ */
        fragmented_data[2] += 1;
        /* atualizamos o ACK */
        fragmented_data[3] += fragmented_data[0].length;

        /* desfragmentamos para enviar */
        array_data[0] = fragmented_data.join(tools.atributes_separator);

        data = array_data[0] + '=TRANSPORT_DONE|PHYSICAL_DIDNT';

        // escreve o terceiro ack no arquivo 
        tools.write_file('transporte+fisica.txt', data);

    }

}

File: modules_perl/blocks.pl
sub hand_shaking_client{

    require '../modules_perl/modules.pl';
    our ( $atribute_separator , $control_separator );

    # pegar parametros passados
	my @s = @_ ;
	my $server_port = $s[0];
	my $server_ip   = $s[1];

    # le conteudo do arquivo com o primeiro ACK
    my $data_file;
    my @array_data_file;
    do{
        # espera a camada de transporte escrever o primeiro ack no arquvivo
        $data_file = read_file('transporte+fisica.txt');
        @array_data_file = split( $control_separator , $data_file );
        print @array_data_file;
    }while(($array_data_file[1] eq 'TRANSPORT_DIDNT|PHYSICAL_DONE') 
                || ($array_data_file[1] eq 'TRANSPORT_DIDNT|PHYSICAL_DIDNT'));

    my @array_message_to_send = split($atribute_separator,$array_data_file[0]);
    # pega a porta de destino
    my $server_port = $array_message_to_send[1];
    # ip -> ESSA DADO DEVE VIR DA CAMADA TCP
    my $server_ip = '172.16.252.75';

    # enviar o primeiro ACK
    my $client_socket = connect_client($server_port,$server_ip);
    my $thread = threads->create(\&send_message,$client_socket,$array_data_file[0]) or die "Erro no envio"; 
    $thread->join();

    # espera o segundo ACK do servidor
    my $receive_bin = <$client_socket>;
    my $receive = sprintf unpack("b*",$receive_bin); 
    $receive = join( "=", $receive , 'TRANSPORT_DIDNT|PHYSICAL_DONE');
    print 'Segundo ack recebido :>' . $receive;
    # escreve o segundo ack no arquivo
    write_file('transporte+fisica.txt',$receive);

    my $shakinghands=1;

    while ($shakinghands) {
        $data_file = read_file('transporte+fisica.txt');
        if (defined $data_file){
            @array_data_file = split( $control_separator , $data_file );
            if ($array_data_file[1] ne 'TRANSPORT_DIDNT|PHYSICAL_DONE'){
                $shakinghands=0;
            }
        }
    }

    # enviar o terceiro ACK
    $thread = threads->create(\&send_message,$client_socket,$array_data_file[0]) or die "Erro no envio"; 
    $thread->join();
    
}

sub hand_shaking_server{

    # pegar parametros passados
	my @s = @_ ;
	my $port = $s[0];
	my $address = $s[1];

    require '../modules_perl/modules.pl';
    our ( $atribute_separator , $control_separator );

    my $socket = create_server($port,$address);
    my $server_socket = $socket->accept() or die "[ERRO CAMADA FISICA] erro no inicio de conexao com o cliente"; # aceita ou nao a conexao com um cliente

    # recebe o primeiro ACK
    my $receive_bin = <$server_socket>;
    my $receive = sprintf pack("b*",$receive_bin); 
    print 'Primeiro ack recebido :>' . $receive;

    # escreve no arquivo o primeiro ACK
    write_file('transporte+fisica.txt',$receive . 'TRANSPORT_DIDNT|PHYSICAL_DONE');

    my $data_file;
    my @array_data_file;
    # fica esperando a camada de transporte ler e escrever o segundo ACK
    do{
        $data_file = read_file('transporte+fisica.txt');
        @array_data_file = split( $control_separator , $data_file );
    }while(($array_data_file[1] eq 'TRANSPORT_DIDNT|PHYSICAL_DONE')
                || ($array_data_file[1] eq 'TRANSPORT_DIDNT|PHYSICAL_DIDNT'));

    # enviar o segundo ACK
    my $thread = threads->create(\&send_message,$server_socket,$array_data_file[0]) or die "Erro no envio"; 
    $thread->join();

    # espera o terceiro ACK
    $receive_bin = <$server_socket>;
    $receive = sprintf pack("b*",$receive_bin); 
    print 'Terceiro ack recebido :>' . $receive;

    # escreve no arquivo o terceiro ACK
    write_file('transporte+fisica.txt',$receive + 'TRANSPORT_DIDNT|PHYSICAL_DONE');

}


## eh necessario retorna um true
1;